src/Pos/TxHistory.hs

```diff
@@ -0,0 +1,241 @@
+deriveAddrHistory
+    -- :: (Monad m, Ssc ssc) => Address -> [Block ssc] -> TxSelectorT m [TxHistoryEntry]
+    :: (Monad m) => Address -> [Block ssc] -> TxSelectorT m [TxHistoryEntry]
+deriveAddrHistory addr chain = identity %= filterUtxoByAddr addr >>
+                               deriveAddrHistoryPartial [] addr chain
+
+deriveAddrHistoryPartial
+    :: (Monad m)
+    => [TxHistoryEntry]
+    -> Address
+    -> [Block ssc]
+    -> TxSelectorT m [TxHistoryEntry]
+deriveAddrHistoryPartial hist addr chain =
+    DL.toList <$> foldrM updateAll (DL.fromList hist) chain
+  where
+    updateAll (Left _) hst = pure hst
+    updateAll (Right blk) hst = do
+        txs <- getRelatedTxs addr $
+                   map (over _1 withHash) (blk ^. blockTxas)
+        let difficulty = blk ^. difficultyL
+            txs' = map (thDifficulty .~ Just difficulty) txs
+        return $ DL.fromList txs' <> hst
+
+----------------------------------------------------------------------------
+-- MonadTxHistory
+----------------------------------------------------------------------------
+
+-- | A class which have methods to get transaction history
+class Monad m => MonadTxHistory m where
+    getTxHistory
+        :: SscHelpersClass ssc
+        => Tagged ssc (Address -> Maybe (HeaderHash, Utxo) -> m TxHistoryAnswer)
+    saveTx :: (TxId, TxAux) -> m ()
+
+    default getTxHistory
+        :: (SscHelpersClass ssc, MonadTrans t, MonadTxHistory m', t m' ~ m)
+        => Tagged ssc (Address -> Maybe (HeaderHash, Utxo) -> m TxHistoryAnswer)
+    getTxHistory = fmap (fmap lift) <$> getTxHistory
+
+    default saveTx :: (MonadTrans t, MonadTxHistory m', t m' ~ m) => (TxId, TxAux) -> m ()
+    saveTx = lift . saveTx
+
+instance MonadTxHistory m => MonadTxHistory (ReaderT r m)
+instance MonadTxHistory m => MonadTxHistory (StateT s m)
+instance MonadTxHistory m => MonadTxHistory (KademliaDHT m)
+instance MonadTxHistory m => MonadTxHistory (PeerStateHolder m)
+instance MonadTxHistory m => MonadTxHistory (NtpSlotting m)
+instance MonadTxHistory m => MonadTxHistory (SlottingHolder m)
+
+deriving instance MonadTxHistory m => MonadTxHistory (PC.ContextHolder ssc m)
+deriving instance MonadTxHistory m => MonadTxHistory (SscHolder ssc m)
+deriving instance MonadTxHistory m => MonadTxHistory (DelegationT m)
+deriving instance MonadTxHistory m => MonadTxHistory (USHolder m)
+
+instance ( MonadDB m
+         , MonadThrow m
+         , WithLogger m
+#ifdef WITH_EXPLORER
+         , MonadSlots m
+#endif
+         , PC.WithNodeContext s m) =>
+         MonadTxHistory (TxpHolder m) where
+    getTxHistory :: forall ssc. SscHelpersClass ssc
+                 => Tagged ssc (Address -> Maybe (HeaderHash, Utxo) -> TxpHolder m TxHistoryAnswer)
+    getTxHistory = Tagged $ \addr mInit -> do
+        tip <- GS.getTip
+
+        let getGenUtxo = filterUtxoByAddr addr . PC.ncGenesisUtxo <$> PC.getNodeContext
+        (bot, genUtxo) <- maybe ((,) <$> GS.getBot <*> getGenUtxo) pure mInit
```

Comment:
You can use [`whenNothingM`](https://github.com/serokell/universum/blob/9edb5d374ec1fb6bb873dacc22a013f1eacd9c67/src/Monad/Maybe.hs#L39) here